compiler/src/dotty/tools/dotc/core/TyperState.scala

```diff
@@ -17,112 +17,93 @@ import scala.annotation.internal.sharable
 }
 
-class TyperState(private val previous: TyperState /* | Null */) {
+class TyperState() {
 
-  Stats.record("typerState")
+  private var myId: Int = _
+  def id: Int = myId
 
-  val id: Int = TyperState.nextId
-  TyperState.nextId += 1
+  private var previous: TyperState = _
 
-  private var myReporter =
-    if (previous == null) new ConsoleReporter() else previous.reporter
+  private var myReporter: Reporter = _
 
   def reporter: Reporter = myReporter
 
   /** A fresh type state with the same constraint as this one and the given reporter */
   def setReporter(reporter: Reporter): this.type = { myReporter = reporter; this }
 
-  private var myConstraint: Constraint =
-    if (previous == null) OrderingConstraint.empty
-    else previous.constraint
+  private var myConstraint: Constraint = _
 
   def constraint: Constraint = myConstraint
   def constraint_=(c: Constraint)(using Context): Unit = {
     if (Config.debugCheckConstraintsClosed && isGlobalCommittable) c.checkClosed()
     myConstraint = c
   }
 
-  /** Reset constraint to `c` and mark current constraint as retracted if it differs from `c` */
-  def resetConstraintTo(c: Constraint): Unit = {
-    if (c `ne` myConstraint) myConstraint.markRetracted()
-    myConstraint = c
-  }
-
-  private val previousConstraint =
-    if (previous == null) constraint else previous.constraint
+  private var previousConstraint: Constraint = _
 
-  private var myIsCommittable = true
+  private var myIsCommittable: Boolean = _
 
   def isCommittable: Boolean = myIsCommittable
 
-  def setCommittable(committable: Boolean): this.type = { this.myIsCommittable = committable; this }
+  def setCommittable(committable: Boolean): this.type =
+    this.myIsCommittable = committable
+    this
 
   def isGlobalCommittable: Boolean =
     isCommittable && (previous == null || previous.isGlobalCommittable)
 
-  private var isShared = false
+  private var isCommitted: Boolean = _
 
-  /** Mark typer state as shared (typically because it is the typer state of
-   *  the creation context of a source definition that potentially still needs
-   *  to be completed). Members of shared typer states are never overwritten in `test`.
-   */
-  def markShared(): Unit = isShared = true
+  /** The set of uninstantiated type variables which have this state as their owning state */
+  private var myOwnedVars: TypeVars = _
+  def ownedVars: TypeVars = myOwnedVars
+  def ownedVars_=(vs: TypeVars): Unit = myOwnedVars = vs
 
-  private var isCommitted = false
+  /** Initializes all fields except reporter, isCommittable, which need to be
+   *  set separately.
+   */
+  private[core] def init(previous: TyperState, constraint: Constraint): this.type =
+    this.myId = TyperState.nextId
+    TyperState.nextId += 1
+    this.previous = previous
+    this.myConstraint = constraint
+    this.previousConstraint = constraint
+    this.myOwnedVars = SimpleIdentitySet.empty
+    this.isCommitted = false
+    this
+
+  def disable() =
```

Comment:
Looks like this is unused currently, I guess it should be called near the end of `explore`.

Reply:
I don't think we need it anymore.